### ISSUE TITLE
Cannot start docker container on ubuntu

### ISSUE THREAD

**Issue by ktruckenmiller** [remie/YouTransfer#118]
Tried on a local ubuntu server and on an EC2 instance but I got the same error when I tried to start the container. 

Recreating youtransfer_web_1...
Cannot start container 6185cd8b7b6e480743563d2df9942c904b1bc53991febc5076ff09aeda9619dc: [8] System error: not a directory

It will create a settings.js FOLDER, and not a file.  It also creates a config.js folder.

**Comment by remie** [comment]
@ktruckenmiller : did you try the latest version (1.0.3)?

**Comment by ktruckenmiller** [comment]
Yeah - I built the latest version a few minutes ago just to make sure. Same thing.

Works fine on OSX though, just can't launch it to a local ubuntu or ec2 ubuntu because of the error

**Comment by mngrant** [comment]
I work with @ktruckenmiller  - we've pulled a few builds over the last 24 hours. Both from GitHub and from the Docker hub. No luck. I assume we're current. 'docker images' shows "latest" on our instances.

**Comment by remie** [comment]
Ok, I'm going to spin up a EC2 instance and see if I can get this to work!

**Comment by remie** [comment]
@ktruckenmiller  / @mngrant : can you share the actual docker run command (incl. parameters)? I'm having some trouble reproducing this error.

**Comment by ktruckenmiller** [comment]
Here's my Yaml for docker-compose
web:
  image: remie/youtransfer
  ports:
    - "80:5000"
  volumes:
    - ./uploads:/opt/youtransfer/uploads
    - ./settings.json:/opt/youtransfer/settings.json
    - ./config.json:/opt/youtransfer/config.json

Or cmd line if you want:

docker run -d -v ./uploads:/opt/youtransfer/uploads -v ./config.json:/opt/youtransfer/config.json -v ./settings.json:/opt/youtransfer/settings.json -p 80:5000 remie/youtransfer

Starting 6ca05adaa1_youtransfer_web_1...
Cannot start container 6ca05adaa1ca8d1f0717c867f2ada6c37b1f6f6a639869526e0427db00337ce2: [8] System error: not a directory

**Comment by mngrant** [comment]
Sure thing

sudo docker run -d -v /mnt/containers/uploads/:/opt/youtransfer/uploads -v /mnt/containers/config.json:/opt/transfer/config.json -v /mnt/containers/settings.json:/opt/youtransfer/settings.json -p 80:5000 remie/youtransfer

**Comment by remie** [comment]
@ktruckenmiller / @mngrant :+1:
Thanks for the extra information, I will continue investigating. The [docker documentation](https://docs.docker.com/userguide/dockervolumes/#mount-a-host-file-as-a-data-volume) tells us that mounting files should work!

**Comment by remie** [comment]
Ok... I've been able to reproduce the error. The issue seems to be caused by the fact that the file does not exist upon initial installation. Apparently, when mounting to a non-existing file, Docker will assume you are mounting a directory. 

I'm going to try and fix this by listing the settings.json file as a mountable volume in the Dockerfile.

**Comment by mngrant** [comment]
Excellent. Thanks for the update.

Worth noting, in our tests we did attempt to touch the files to create empty files first. That did not work. On run the directories (same name) are created next to the files we touched.

**Comment by remie** [comment]
I managed to get it to work with

````
sudo docker run -d -v /mnt/containers/uploads/:/opt/youtransfer/uploads -v /mnt/containers/config.json:/opt/youtransfer/config.json -p 5000:5000 --name youtransfer remie/youtransfer
````

but this still excludes the settings.json file. I will continue to look for a solution!

**Comment by ktruckenmiller** [comment]
Still the same behavior for me. Now it just created a config.js/ folder

**Comment by remie** [comment]
For some strange reason I cannot reproduce it anymore. I'm using Docker version 1.8.2, build 0a8c2e3 on Amazon EC2 with AMI "ubuntu-trusty-14.04-amd64-server-20150325 (ami-47a23a30)" in Ireland region (eu-west-1).

Upon boot of the instance I run `curl -sSL https://get.docker.com | sh` and create the config.json file (with default values), a settings.json (with empty object {}) and a uploads/ folder in /mnt/containers.

Finally start the Docker image with 
````
sudo docker run -d -v /mnt/containers/uploads/:/opt/youtransfer/uploads -v /mnt/containers/config.json:/opt/youtransfer/config.json -p 5000:5000 --name youtransfer remie/youtransfer
```
and it just works.

@ktruckenmiller / @mngrant : is it possible for you guys to follow these steps and see if you can get this working?

**Comment by mngrant** [comment]
Doing that works!

I think some documentation about using the default config from the github and an empty object settings file would be helpful.

**Comment by remie** [comment]
@mngrant : yeah, but that seems like a bit of a hassle... it should just work, especially in regard to the default config.json file that ships with the NPM package. I'm glad we've found a workaround for you guys but I will keep this issue open and tinker with it tomorrow to see if I can fix this.

**Comment by klingtnet** [comment]
When you mount `{config,settings}.json` as docker volumes, then you have to **make sure that the files exist on your docker host**, otherwise docker assumes that they are directories and will create them for you.

**Comment by remie** [comment]
@klingtnet : thanks for confirming my suspicions! I guess I've learned it the hard way :) I'm working on a fix now which will simplify the docker configuration. See as294d9 for more details!

**Comment by remie** [comment]
@klingtnet : the link to the data-container does not work. Can you elaborate on this?

**Comment by klingtnet** [comment]
I fixed the link and should check my clipboard next time :disappointed: ...

**Comment by remie** [comment]
:) I also found [this article](https://medium.com/@ramangupta/why-docker-data-containers-are-good-589b3c6c749e), which helps understand it. As I'm now moving the config files into folder for the docker container, data-containers become easier to use. I will update the documentation with links to how to implement this pattern.

**Comment by klingtnet** [comment]
There are some pros and cons using data containers:

Cons:
- editing files is a little bit more work because they are saved inside the container. My workflow looks like this: `docker run --volumes-from some-data-container klingtdotnet/vim`

Pros:


- create the data container `docker create --name youtransfer-data --volume /path/to/config --volume /path/to/upload/data remie/youtransfer`
- run YouTransfer `docker run --volumes-from youtransfer-data -p 80:5000 remie/youtransfer`

**Comment by klingtnet** [comment]
There are some pros and cons using the data containers approach:

Cons:
- editing files is a little bit more work because they are saved inside the container. My workflow looks like this: `docker run --rm --volumes-from some-data-container -it klingtdotnet/vim`

Pros:
- backup is easy
- there are no permission problems, because both, the data-container and the app-container, share the same users (that's why you should use the same image for both)

- create the data container `docker create --name youtransfer-data --volume /path/to/config --volume /path/to/upload/data remie/youtransfer`
- run YouTransfer `docker run --volumes-from youtransfer-data -p 80:5000 remie/youtransfer`

**Comment by remie** [comment]
@ktruckenmiller / @mngrant : i've just released version 1.0.5 which includes the fix for your issue. I still need to update the documentation, but it basically involves mounting a config folder instead of the individual files. You can now use:

````
sudo mkdir /mnt/containers/config
sudo mkdir /mnt/containers/uploads

sudo docker run -d -v /mnt/containers/uploads/:/opt/youtransfer/uploads -v /mnt/containers/config/:/opt/youtransfer/config/ -p 80:5000 --name youtransfer remie/youtransfer
````
